compiler-plugin/src/main/scala/scalapb/compiler/ProtobufGenerator.scala

```diff
@@ -50,33 +50,41 @@ class ProtobufGenerator(
       }
       .add(s"object $name extends ${e.companionExtends.mkString(" with ")} {")
       .indent
+      .seq(e.traitAnnotationList)
       .add(s"sealed trait ${e.recognizedEnum.nameSymbol} extends $name")
       .add(s"implicit def enumCompanion: _root_.scalapb.GeneratedEnumCompanion[$name] = this")
+      .newline
       .print(e.getValues.asScala) { case (p, v) =>
         p.call(generateScalaDoc(v))
-          .add(s"""@SerialVersionUID(0L)${if (v.getOptions.getDeprecated) {
-            " " + ProtobufGenerator.deprecatedAnnotation
-          } else ""}
-                  |case object ${v.scalaName.asSymbol} extends ${v.valueExtends
+          .add("@SerialVersionUID(0L)")
+          .seq(e.valueAnnotationList)
+          .seq(v.annotationList)
+          .add(s"""case object ${v.scalaName.asSymbol} extends ${v.valueExtends
             .mkString(" with ")} {
                   |  val index = ${v.getIndex}
                   |  val name = "${v.getName}"
                   |  override def ${v.isName}: _root_.scala.Boolean = true
                   |}
                   |""".stripMargin)
       }
-      .add(s"""@SerialVersionUID(0L)
-              |final case class Unrecognized(unrecognizedValue: _root_.scala.Int) extends $name(unrecognizedValue) with _root_.scalapb.UnrecognizedEnum
-              |
-              |lazy val values = scala.collection.immutable.Seq(${e.getValues.asScala
-        .map(_.scalaName.asSymbol)
-        .mkString(", ")})
-              |def fromValue(__value: _root_.scala.Int): $name = __value match {""".stripMargin)
+      .add("@SerialVersionUID(0L)")
+      .seq(e.unrecognizedAnnotationList)
+      .add(
+        s"""final case class Unrecognized private[$name](unrecognizedValue: _root_.scala.Int) extends $name(unrecognizedValue) with _root_.scalapb.UnrecognizedEnum
+           |object Unrecognized {
+           |  @deprecated("Could have lead to issues before. Use $name.fromValue instead. This might be private in the future.")
+           |  def apply(unrecognizedValue: _root_.scala.Int): $name = fromValue(unrecognizedValue) 
```

Comment:
That's a great idea, but unfortunately it breaks binary compatibility (see the mima test). I suggest that the changes around this (deprecation and making the constructor private) are split to a different PR which we can target for next stable (0.12 or 1.0)